rudimentary aspect ratios

Gerschel opened the pull request:
Adds a ui setting where a user can specify their
preferred aspect ratios.  

Adds a ui setting where a user can specify the base
dimension (the one to keep the same, affects other).

~~Uses this to generate the buttons for txt2img &
img2img, with element ids.~~

To keep the user experience intuitive, the ratio uses colon format ":" to separate width and height.
This value is used as the ~~button~~ dropdown value, parsed as ints for the ratio value, ~~and used in the element id.~~

I did not update the stylesheets. I think there may need to be an update for ~~these buttons~~ this dropdown. ~~perhaps using `"button[id*=ratio]"` to target them, maybe even the row, so it can become size/ratio row.~~


**Environment this was tested in**

 - OS: Windows
 - Browser: chrome, firefox
 - Graphics card: not relevant

![image](https://user-images.githubusercontent.com/9631031/216793001-dd59a3ef-86ae-4859-86f0-2d06cab337c8.png)

Added user setting for specifying dimension to keep, so the other adjusts.
![image](https://user-images.githubusercontent.com/9631031/216795043-43c61f13-f37e-4045-8fe8-0399bf1f4e0b.png)

![image](https://user-images.githubusercontent.com/9631031/216796183-9f5cb64f-4ac1-4451-83dd-065fc16545ee.png)

Gerschel committed:
rudimentary aspect ratios

Adds a ui setting where a user can specify their
preferred aspect ratios.
Uses this to generate the buttons for txt2img &
img2img, with element ids.
To keep the user experience intuitive, the ratio uses colon format
":" to separate width and height.
This value is used as the button value, parsed as ints for the ratio
value, and used in the element id.

Gerschel committed:
fixed aspect ratio math

aspect ratio change is based on larger dimension

w-e-w commented:
I think it makes more sense to make it into a drop down menu from a small button
![ui concept](https://user-images.githubusercontent.com/40751091/216794354-a3d3d042-e33d-4d35-800c-f5e615629c9b.png)
(this is photoshopped)
or similer to `Sampling method` dorp down menu

Gerschel committed:
added dimension setting for user to specify base

Modified aspect ratio resize to use user preferred dimension
Included setting next to aspect ratio inputs.

Gerschel commented:
> I think it makes more sense to make it into a drop down menu from a small button ![ui concept](https://user-images.githubusercontent.com/40751091/216794354-a3d3d042-e33d-4d35-800c-f5e615629c9b.png) (this is photoshopped) or similer to `Sampling method` dorp down menu

I agree. I'll see what I can do. This is exactly why I made the pull request before doing the final touches, feedback.

Gerschel committed:
change from buttons to a single dropdown

Gerschel commented:
Removed buttons, changed to a single dropdown.  
This removes the element id's for the buttons as well.  
Dropdown has element id "txt2img_ratio" and "img2img_ratio" respectively.  

No styles have been applied yet.

![image](https://user-images.githubusercontent.com/9631031/216796183-9f5cb64f-4ac1-4451-83dd-065fc16545ee.png)

Gerschel commented:
I'm going to convert this to a class and implement a lock/unlock.  
Where, if you adjust either width or height, the other slider will move according to the ratio selected, if the checkbox is checked.  
Something like this:
![image](https://user-images.githubusercontent.com/9631031/216799407-e94a3abf-f092-49d9-820b-8e78ef0a216e.png)

Gerschel committed:
added aspect ratio lock; dependent on base option

Gerschel commented:
Added aspect ratio lock. It's dependent on base dimension in setting. Move the slider on the base dimension, width or height, it updates the other value. But it does not work the other way around, because that causes a circular change event, so it's scoped to work only on base dimension.  

![image](https://user-images.githubusercontent.com/9631031/216802266-0e638287-9b31-41f6-bf0b-c1c55723a0fd.png)

w-e-w commented:
### some other feed back

1 
![image](https://user-images.githubusercontent.com/9631031/216795043-43c61f13-f37e-4045-8fe8-0399bf1f4e0b.png)
text input is not user-friendly [Radio](https://gradio.app/docs/#radio-header) is better

2
instead of a checkbox enable
you could make the 1 first option in the drop-down menu the `disabled option`

3
![image](https://user-images.githubusercontent.com/40751091/216803500-0f3f1111-7f61-423d-831e-3c25e5e429a6.png)
because the way that the slider constantly flashes is extremely annoying

also due to the update is done on the service side
if you drag the slider too fast the newest update for not be lost, making experience "not smooth"
can caused the fire aspect ratio to be incorrect

I'm no gradio expert but it would be better if this on client side, maybe is not possible with gradio 

### other note
I'm playing around with wish different layouts such as using `Accordion`
![image](https://user-images.githubusercontent.com/40751091/216803334-71d8c6b7-2beb-4a3d-b3c2-3e6089c2cc09.png)
![image](https://user-images.githubusercontent.com/40751091/216803335-49aff03f-c7a6-404c-a1de-3e3e9676ca9a.png)
or put it with cfg scale
![image](https://user-images.githubusercontent.com/40751091/216804207-2de26340-f109-449a-9760-6a409f6de70e.png)
(I don't think AUTO will like this position)

just in case you don't know there is a `--ui-debug-mode` for quick UI loading

Gerschel commented:
@w-e-w  
I do have a little library I wrote in javascript that handles every type of element for this webui, client side. 
I had written a PresetManager version 2 that allowed you to save, load, apply as stackables and non stackabels, any setting on any tab. I took it down because I got tired of updating it while there were a lot of changes. I considered bringing it back in the future, I didn't want to give it away, because I had over 100 hundred hours doing a rewrite from version 1 to 2, and others have stolen my work and submitted as their own with a few other projects (I just wanted the thunder for my hard work). https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/6808

So I know for a fact this can be done.

Here's an example for the sliders, you can use the methods `getVal` and `setVal`. The class takes in an element as an argument.  

You can try dropping this in the browsers devtools console to load the class.

```javascript
class PresetManagerSliderComponent {
    constructor(element) {
        this.element = element;
        this.childNumField = this.element.querySelector('input[type=number]');
        this.childRangeField = this.element.querySelector('input[type=range]');
    }
    getVal() {
        return this.childNumField.value;
    }
    updateVal(text) {
        this.childNumField.value = text;
        this.childRangeField.value = text;
    }
    eventHandler() {
        this.element.dispatchEvent(new Event("input"));
    }
    setVal(text) {
        this.updateVal(text);
        this.eventHandler();
    }
}
```

Then to instance it you'll use the `new` keyword:
```javascript
let test = new PresetManagerSliderComponent(gradioApp().querySelector("#txt2img_height"))
```  

Then you can use the get and set methods like you would expect.  
```javascript
test.setVal(42)
```

Gerschel commented:
Is there a way I can pause this pull request until I make the changes? Maybe close it then reopen it?

Gerschel commented:
I need to close this. Rollback the master, branch it, so I can submit a different pull request. Then when I get back to this I can try to resubmit it from a branch.

Gerschel commented:
Created new draft from a branch, you will have to clear the 'aspect_ratios' and 'aspect_ratios_base' from config.json if you tested this version, before testing the new draft.  
https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/7601

Gerschel commented:
A big difference, now that I make the change using javascript, I don't need the checkbox, and I moved the disable as an option in the dropdown. Also tossed the base dimension requirement, it was necessary before, because it caused an circular issue when it used gradio's default event handlers. Now you can adjust either and it reflects after letting go, to avoid the circular issue.
